### ISSUE TITLE
Forecaster interface: getting in-sample forecasts, getting residuals

### ISSUE THREAD

**Issue by AngelPone** [alan-turing-institute/sktime#1105]
**Is your feature request related to a problem? Please describe.**
I think there should be a method or property to obtain in-sample fitted value or residuals, which are important in many forecasting methods, such as forecast combination.

**Describe the solution you'd like**

Add a method or property to `BaseForecaster` and it should be implemented for every forecaster.

**Additional context**
A traditional forecast combination method based on variance of in-sample residuals.

**Comment by fkiraly** [comment]
Ok, I think there are two requests in there, and a proposed way of addressing these.

The two requests: residuals and in-sample values.

Regarding residuals, I fully agree that there should be an easy way to return them!
But, I don't think this is a concern of the `BaseForecaster`.

There also already is a way to obtain residuals from a forecaster, it's the `Detrender` - simply detrend with the forecaster you would like to get residuals of - it's a transformer, and the output gives you the residuals.
Perhaps we should say this in the forecasting tutorial, since it's not obvious from the name "detrender"?

Regarding in-sample fitted values: Is this not what you would get when you run `predict` at the values that have already been seen? Simply pass a fixed forecasting horizon with the indices that you've observed.

`forecaster.fit(y).predict(fh=y.index)` or `forecaster.fit(y, fh=y.index).predict()`. Does this work for you?

What I'm wondering, should there perhaps be a `fit_predict(y)`, which, if `fh` is not passed, defaults to in-sample forecasts, i.e., `fh=y.index`?

Feedback on whether it is what you need, or whether it's "obvious" would be appreciated.

**Comment by fkiraly** [comment]
for forecast combination, you may like to look at the `HedgeExpertEnsemble`, it may do what you want, i.e., re-weight depending on errors or forecast residuals.

Would be nice to hear from you whether you think that's what you need or not, and whether it should be made clearer in the tutorial.

**Comment by RNKuhns** [comment]
@fkiraly and @AngelPone outside of the forecast combination case, I could see it being useful to be able to expect the residuals from a model. For example, to see if they have a clear pattern (suggesting the model doesn't fit well) or run a statistical test on them (e.g. are they normally distributed in the case of models like Arima that make this assumption).

Adding a convenience method would be pretty straight-forward -- something like:

```python
class BaseForecaster(BaseEstimator):
    def insample_residuals(self):
        self.check_is_fitted()
        if self.get_tag("requires-fh-in-fit", False):
             # Not sure if we should use `_fit` or `fit` here
             # Maybe `_fit` since it doesn't override the stored `fh`?
            return self._fit(self._y, fh=self._y.index).predict()
        else:
            return self.predict(fh=self._y.index)
```

**Comment by aiwalter** [comment]
Regarding residuals:
from a [discussion on Slack](https://sktime-group.slack.com/archives/CQVH0NJ1M/p1628935128001400) with @ngupta23 : As some models e.g. from `statsmodels` have residuals internally calculated, the best would be to have a default function `def _get_residuals(self)` or sth in the parent `fit()` after` _fit()` was called that calculates residuals based on in-sample predictions. So we could overwrite this function in the child class for some models that provide this internally. So the parent class would get an additional attribute `residuals_`. Not sure if this should be optionally or always calculated. The reason for optionally would be that for large data sets, we maybe dont want to store the residuals in `self`.
What does everybody think?

**Comment by fkiraly** [comment]
alternative suggestion: a `ResidualsTrafo` transformer that wrape a forecaster? That way, it can be used as a transformer in pipelines.

**Comment by mloning** [comment]
I think you can just run:
```python
transformer = Detrender(MyForecaster())
residuals = transformer.fit_transform(y)
```

**Comment by fkiraly** [comment]
oh yes, we already have the `ResidualsTrafo` and it's called `Detrender`.

Perhaps we should introduce an alias `Residuals` and/or add a "computing residuals" section to the tutorial?

**Comment by ngupta23** [comment]
So just to clarify, `Detrender` will give the true in-sample residuals (in-sample prediction MINUS in-sample actual)?

**Comment by fkiraly** [comment]
I think so, yes.

**Comment by ngupta23** [comment]
> Perhaps we should introduce an alias Residuals and/or add a "computing residuals" section to the tutorial?

Yes, that will be a useful addition (from a clarity perspective) even though it will be an alias.

**Comment by ngupta23** [comment]
> 
> 
> I think you can just run:
> 
> ```python
> transformer = Detrender(MyForecaster())
> residuals = transformer.fit_transform(y)
> ```

Hi @mloning , Would it be possible to store the residual information in the forecaster object when fit is called for it? This would be more efficient since usually, the residuals are called only after fitting the forecaster (for fit assessment). Having to call another fit after that in `transformer.fit_transform` would be doubling the fit effort in terms of time. 

```python
forecaster = ExponentialSmoothing(sp=12, seasonal="mul", trend="add")
forecaster.fit(y)  ## store residuals in the forecaster object when fit is called.
```

If we do not want to enforce this (due to storage overhead), then maybe adding an argument to fit called `save_residuals: bool=False` would be nice to have.

```python
forecaster = ExponentialSmoothing(sp=12, seasonal="mul", trend="add")
forecaster.fit(y)                       ## Does not store residuals in the forecaster object when fit is called.
forecaster.fit(y, save_residuals=True)  ## Stores residuals in the forecaster object when fit is called.
```

**Comment by fkiraly** [comment]
@ngupta23, why would there be an additional call to `fit`? The `Detrender` calls `fit_transform` of the wrapped forecaster, which calls `fit`, and it happens only once?

**Comment by ngupta23** [comment]
@fkiraly, usually I would fit the model first and only then evaluate (optional). If someone already has a fitted model in the workflow, then having to wrap it in a `Detrender` to get the residuals (using fit_transform) would force another fit on that and would duplicate the effort.

**Comment by ngupta23** [comment]
> Further, I don't think your suggestion works, since for obtaining residuals you need to not only fit but also predict.

@fkiraly maybe it was not explicitly stated, but in my suggestion, we would need to do insample prediction when save_residuals=True, somethig like what is done inside the detrender currently [Link](https://github.com/alan-turing-institute/sktime/blob/ebd8bbf088276240bb924fca10be0624d8879dff/sktime/transformations/series/detrend/_detrend.py#L141)

**Comment by mloning** [comment]
If you don't want to fit again, you could also do:
```python
f.fit(y)
residuals = y - f.predict(ForecastingHorizon(y_train.index, is_relative=False)
```

**Comment by ngupta23** [comment]
Hi @mloning ,

It is more than just computing the residuals. If we are working with multiple models, we want to keep track of the residuals for each model. The model object is the best place to store this in my opinion. So, I like your idea of adding a `residuals_` property to the forecaster class. 

I can also see why computing the residuals inside the `fit` method could be questioned. Maybe we can compromise and add a method called `compute_residuals` that can be called right after the `fit` by the end user if they choose to do so. This method would compute the residuals and store it in the `residuals_` property. 

Do you think this would be an acceptable compromise to implement?

**Comment by mloning** [comment]
I'd actually prefer `compute_residuals` over `residuals_`, that could be added to the `BaseForecaster`

**Comment by fkiraly** [comment]
How about `predict_residuals`, since it's a "prediction-like" method?

**Comment by fkiraly** [comment]
In any case, we need to be very precise what the methods are doing - in-sample, out-of-sample, etc.

I'd be in favour of `fit_predict` and `fit_predict_residuals` for in-sample prediction and in-sample residuals.
For out-of-sample residuals you could have `predict_residuals`?

**Comment by mloning** [comment]
I'd prefer keeping it simple and only provide a method for in-sample residuals, at least for now. Don't have preference over `compute_residuals` or `predict_residuals`.

**Comment by ngupta23** [comment]
So are we good with implementing `compute_residuals` or `predict_residuals` for in-sample residuals and storing the residual values in the model object?

**Comment by mloning** [comment]
Do you need to store them? I believe that if we store them, we need to handle that state properly, making sure in sync with the training data. Implementing `compute_residuals` as some form of `y_train - f.predict(ForecastingHorizon(y_train.index, is_relative=False)` sounds good to me.

**Comment by fkiraly** [comment]
yes, but let's write a short design doc before implementing anything, it's a core API change, so strictly speaking it needs a STEP.
At least, you should write down docstring and method signature.

I'd do `predict_residuals(y: Series) -> Series`, i.e., the `fh` is directly obtained from `y`.

**Comment by fkiraly** [comment]
I'd also add a `fit_predict(y: Series) -> Series`.

**Comment by ngupta23** [comment]
> Do you need to store them? 

@mloning , I think we should store this in the model object because when we work with multiple models for the same data, then storing this in the model makes most sense.

> I believe that if we store them, we need to handle that state properly, making sure it's in sync with the training data.

Yes, that is why my [original recommendation](https://github.com/alan-turing-institute/sktime/issues/1105#issuecomment-913121620) was to do this when doing fit (with an argument that has to be explicitly set to True to store this so it does not impact the default functionality).

**Comment by ngupta23** [comment]
@mloning @fkiraly @aiwalter Do we need to discuss this further? Any help in closing this would be appreciated as we need this functionality for `pycaret` development as well.

**Comment by fkiraly** [comment]
@ngupta23, I think the end state is not clear enough, e.g., what the design would be. See discussion above.

* you say `fit` would be called twice if we want to compute residuals using the current interface. I don't see this, please explain.
* among the different options, we haven't settled on one.
* it's also not clear who would implement this.

**Comment by fkiraly** [comment]
@ngupta23, @AngelPone, we could add the functionality with our extension of the `predict` interface, see here?
https://github.com/alan-turing-institute/sktime/pull/1421

**Comment by ngupta23** [comment]
@fkiraly, @mloning   Sorry for the delayed response. I think there are a few requirements that I think came from this discussion:

1. residuals should be stored in the model object. Reason: It helps when we are building multiple models for the same data set and want to compare the results at the end. Without this, the onus will be on the downstream users of `sktime` to keep track of this separately.  
2. If we keep track of the residuals in the model, we need to track there state properly. As per @mloning 
> we need to handle that state (training data used for residuals, etc.) properly, making sure it's in sync with the training data.

I think to satisfy these requirements, I would like to discuss my original proposal of doing this in the `fit` method

```python
forecaster = ExponentialSmoothing(sp=12, seasonal="mul", trend="add")
forecaster.fit(y)                       ## Does not store residuals in the forecaster object when fit is called.
forecaster.fit(y, save_residuals=True)  ## Stores residuals in the forecaster object when fit is called.
```

Fit would look something like this then

```python
def fit(self, y, X=None, fh=None, save_residuals: bool=False):
    # existing code ... fit happens here

    # new addition
    self.residuals = None
    if save_residuals:
        self.residuals = y - self.predict(ForecastingHorizon(y.index, is_relative=False)
```
Let me know what you think?

**Comment by fkiraly** [comment]
@ngupta23, I don't think it's a good idea to:

* make changes to the core interface
* add the "store residuals" concern to `fit`, because `fit` doesn't compute any predictions (which are needed for residuals)

I also think you should clearly verbalize your requirements before presenting a solution.
"residuals should be stored in the model object" is not a requirement - for one, the residuals being stored is not the "ultimate goal". Further, it's vague, since you don't say which residuals you need or want.

From your discussion, I gather you mean in-sample residuals for the same index that the model is fitted to?

Above you say that you would need to call `fit` twice, but your actual code addition does not seem to make a difference between (a) just calling `fit` and specifying an extra argument to it, or (b) calling `fit` and then `predict` on `y.index`.

I don't see why we would need to make major API changes, like adding a `save_residuals` argument, if all you would need to do is

```
y- self.fit(y, fh=fh).predict(y.index)
```

I.e., you propose to add a second, non-obvious way to do things, which does not seem to be that much quicker anyway.

As said, I'd be happy to define the above line as a shorthand `fit_predict_residuals` or similar, but you were not happy with it since you say you would "call `fit` twice", which I don't understand and you haven't explained.

**Comment by ngupta23** [comment]
@fkiraly 

The title of the issue is referring to in-sample forecasts and residuals so I thought that was implicit. But maybe not. So yes, to clarify, I was indeed referring to in-sample residuals.  

The reason why I said it would be calling `fit` twice was in particular reference to this comment ([Link](https://github.com/alan-turing-institute/sktime/issues/1105#issuecomment-898961433)). Usually, one would fit the model first and then compute residuals

```python
# Step 1: Fit the model
model.fit(y)

# Step 2: Compute residuals (in-sample)
transformer = Detrender(model)
residuals = transformer.fit_transform(y)
```

The above flow would call fit inside the fit_transform, hence the call to fit twice (once in the model.fit() call and then in the transformer.fit_transform() call.

But then @mloning clarified ([Link](https://github.com/alan-turing-institute/sktime/issues/1105#issuecomment-913188215)) that we do not need to call this twice if we do something like this which is absolutely fine with me since it is not calling fit twice. 

```python
model.fit(y)
residuals = y - model.predict(ForecastingHorizon(y.index, is_relative=False)
```

If you want to encapsulate this whole thing in what you called `fit_predict_residuals_insample(y)` ([Link](https://github.com/alan-turing-institute/sktime/issues/1105#issuecomment-927260222)) that is also OK with me.

The only question that remains from my side now is whether the residuals would be saved inside the model object or simply returned to the user where they would have to manage it themselves?

The reason for the above is that in `pycaret`, we can build multiple models first and then check goodness of fit later. If the residuals are not stored in the model objects, then we (and any other developers using `sktime` would need to manage these by themselves). That is why I feel storing them inside the model object is better since it avoids replication of work from multiple downstream users of `sktime`. I don't know if that answers your questions. Feel free to let me know if you need more clarification.

**Comment by fkiraly** [comment]
Ok, we're getting closer to the problem here. 
I think you are mistaken about how `scikit-learn` like object oriented interfaces work, and that misunderstanding causes some incorrect reasoning down the line about what new interfaces we should add.

More precisely: your code

```
# model definition
model = Sth(params)

# Step 1: Fit the model
model.fit(y)

# Step 2: Compute residuals (in-sample)
transformer = Detrender(model)
residuals = transformer.fit_transform(y)
```

is incorrect, that's not how you would use the `Detrender`! Have a look at the forecasting tutorial.

The complete code to obtain residuals is

```
# model definition
model = Sth(params)

# compute residuals
transformer = Detrender(model)
residuals = transformer.fit_transform(y)
```

It calls the model's `fit` exactly once, so no double fitting is carried out.

I'd suggest you look at how `sklearn` and `sktime` composites work - *first* you construct them without fitting the components. *then* you can call the result with a single call to `fit` etc.
In a `scikit-learn` pipeline, you also don't fit the components individually before you plug them into the pipeline.
This is notably different from the "classical" design where you would have to control the data ingestion in each individual component.

Does this explanation help?

**Comment by mloning** [comment]
I see your point @ngupta23. 

As we've discussed above, the main problem I see is that storing residuals in the model requires us to ensure that the state of the residuals attribute is consistent with the state of other model attributes (input data, etc). To me, this would add some complexity at this stage which I'd like to avoid.  

So I'm in favour of adding a `compute_residuals` method for in-sample residuals which returns: 
`self._y - model.predict(ForecastingHorizon(self._y.index, is_relative=False)`. You could then handle residuals from different fitted models separately in your code similar to how you handle out-of-sample predictions and performance metrics for different fitted models.

**Comment by ngupta23** [comment]
Hi @mloning I went back and looked at `sklearn` and they also do not store the residuals in the model object so I am ok with your recommendation of providing the method to compute this as an alternate. I have temporarily added this to `pycaret` [[Link](https://github.com/pycaret/pycaret/blob/e37dc52425ebef089c77d5b8a8fe4d3be242f940/pycaret/internal/pycaret_experiment/time_series_experiment.py#L3497)]. I can switch to your method when it is available.

(Just my opinion) I do not necessarily like the idea of the Detrender Wrapper since the primary task of a forecaster is to forecast. Residual analysis is a post processing step to check goodness of fit. If we wrap the forecaster in a Detrender, then yes, we will get the residuals with a single fit call, but to get the model predictions, we may have to dig deeper into the Detrender object to do so. This seems a little round about to me and breaks many model development flows where models are fit first and then residual analysis is performed.

**Comment by fkiraly** [comment]
>I do not necessarily like the idea of the Detrender Wrapper since the primary task of a forecaster is to forecast. 

I don't get why this seems like a problem - the detrender is a transformer, not a forecaster.

>Residual analysis is a post processing step to check goodness of fit.

And that's why it is a concern that sits with the evaluation machinery, not primarily with the forecaster.
The design principle here is "separation of concerns" or "decoupling".
You wouldn't want to attach everything that you can potentially do with results of a forecaster, to the forecaster - or the results themselves, that wouldn't be the sleek and lightweight design we are aiming for.

I know that `pycaret` follows more of a "god function" design around pre-defined workflows, but in `sktime` we are trying to stay modular and flexible... so packages like yours can get from `sktime` what they want easily.

>but to get the model predictions, we may have to dig deeper into the Detrender object to do so

That I think is a valid argument - if you want an easy way to get predictions *and* residuals.

I still think that generic evaluation functionality and residual analysis needs to be in a place different from the forecaster - but residuals are not a big thing, assuming you want them retrieved and not stored in the object.

**Comment by ngupta23** [comment]
Hi @fkiraly , Understood your concerns. I am OK not saving the residuals in the forecaster and I think there is a way forward here. I have added the code from @mloning to `pycaret` for now. If a forecaster method to compute residuals (returned not stored) can be provded, we will use that, else we will continue using the code in `pycaret` as it stands now.

Thanks!
